rfcs/0014-dedicated-serving-cluster.md

```diff
@@ -0,0 +1,96 @@
+
+### Read Checkpointed Snapshots
+
+In the first stage, we can simply boot up several compute nodes without scheduling any streaming tasks on them. The meta service pushes the new storage versions to the compute nodes as usual, therefore, whenever a new storage version generated, the compute node will be informed.
+
+In the catalog, beside the existing `vnode -> parallel unit` mapping, we have to maintain another mapping of serving cluster for each table & materialized view. Below is an example, where “PU” refers to Parallel Unit. Apparently, this mapping should be updated once nodes joined or leaved the serving cluster. 
+
+```jsx
+Main Cluster: 
+- PU#1: vnodes [0, 1]
+- PU#2: vnodes [2, 3]
+- PU#3: vnodes [4, 5]
+
+Serving Cluster: 
+- PU#1: vnodes [0, 1, 2]
+- PU#2: vnodes [3, 4, 5]
+```
+
+High availability can be achieved by allocating multiple Parallel Units for each vnode, where one of them is primary and the other(s) are stand-by. Once the primary replica is unavaialble, the stand-by replica will immediately be picked by frontend as new route. This provides us both high-availability and cache-locality. For example, given 3 nodes and `replicas = 2`,
+
+```jsx
+Serving Cluster: 
+- PU#1: vnodes [0, 1, 2, 4]
+- PU#2: vnodes [3, 4, 5, 1]
+- PU#3: vnodes [0, 2, 3, 5]
+```
+
+### Read Non-Checkpointed Snapshots
+
+In the current design, non-checkpointing barriers are much more frequent than checkpointing barriers, so it provided both freshness and lower IO cost. However, these non-checkpointed updates are only kept in memory, more specifically, in the `StateStore` of the `MaterializedExecutor` with corresponding vnode.
+
+#### Approach 1. Read non-checkpoint writes from primary cluster (Preferred)
+
+> Suggested by @BowenXiao1999 and @hzxa21
+
+Inform the serving cluster with the vnode mapping of primary streaming cluster. Once the servering cluster find it lack of the data of epoch being requested (the snapshot epoch to read is piggy-backed in frontend's RPC), it can send an RPC to fetch the changes (i.e. immutable MemTables) from the corresponding streaming CN.
+
+#### Approach 2. Read non-checkpoint writes from S3
+
+Here we propose to **store** these barrier-level changes of non-checkpoint barriers to shared storage **temporarily**. These immutable MemTable SSTs will only be temporarily placed in L0, and these SST files can be dropped as long as next checkpoint completes.
+
+To make it simpler, we can reuse the concept “compaction group”. A compaction group can be tagged as checkpointing on every barrier. Table_ids in such compaction groups will upload its immutable MemTable on non-checkpointing barriers, and a “serving-only” version will be created.
+
+Note that serving-only versions will not affect the recovery process. On recovery, all nodes should roll back the latest completed checkpoint like before.
+
+#### Approach 3. Contantly streaming changes from primary cluster
```

Comment:
Vote for approach 3, it's similar to share arrangement

Reply:
In our previous design, the incremental updates to share arrangements will be broadcasted to different actors. The behavior is what we expected here.